### ISSUE TITLE
Parallel trades

### ISSUE THREAD

**Issue by davidjirovec** [butor/blackbird#398]
Are there plans for implementing multiple parallel trades? Now with multiple exchanges supporting short selling, Blackbird does not have to wait for market exit to enter market with two different exchanges.

**Comment by butor** [comment]
Yes. It will be specially useful when we are "stuck" in a spread that doesn't want to close. With parallel trades we could still trade other opportunities. It seems to be a major change to the codebase though.

**Comment by CCKRK** [comment]
I don't even know where to start to get the program to run trades simultaneously but I'm currently trying out a few different ways to get BlackBird to remember its orderIds once positions are entered. Once I've got that, altering the "getActivePosition" members (and a few considerations for balance calcs) I will at least be able to run separate instances of the program on the same exchanges with minimal interference. 

Right now I'm storing them as strings in the result struct which may not be very performant.

**Comment by CCKRK** [comment]
@butor so I've gotten my orderId idea to work -- I think it will be an important improvement especially if we wanted to implement order canceling/resubmitting in ref to #386. I could probably run multiple instances of blackbird now in separate containers (because of the result.txt and db files..) for parallelism but this is less than ideal.

Any high level ideas on what it would take to accomplish parallel trades? Could we lose the inmarket/!inmarket condition in main.cpp, construct a new vector that mirrors btcVec with the in market exchanges once an entry is found, and loop on that vector right after the main arbitrage opportunity loop(s)?

**Comment by butor** [comment]
@CCKRK When I started developing Blackbird, my strategy was to go all in on one opportunity, wait for it to close, then move on to the next one. This was feasible because the spreads, back in 2014, were wide and short-term (couple of minutes / hours) due to high volatility and high market inefficiency at that time. It was possible to capture many opportunities a day.

As we've seen this is no longer the case, and the opportunities can last days. So it's a good idea to add parallelism. But it's a major work because, as you've pointed out, a lot of variables (such as `inMarket`, `limPriceLong`, `limPriceShort`, but also `Result` structure) will need to be in arrays and the overall balance management will be more complex. I can definitely help with that.

Note: another option would be to increase the trade frequency and try to capture narrower spreads that should last a few minutes or even seconds. But I usually see too much network latency to implement something high-frequency like that.

**Comment by CCKRK** [comment]
I've essentially got it working doing something similar to what I described -- basically just two Result struct vectors with persistence of active positions (in case we exit with positions open) using the db instead of the result.txt. Just pops them in and out when checkEntry/checkExit finished. Makes the log kind of messy 
```
 [02/12/2018 13:58:15 ]
   Kraken: 	8,790.00 / 8,791.00
   Bittrex: 	8,775.00 / 8,793.00
   GDAX: 	8,769.98 / 8,769.99
   Binance: 	8,785.12 / 8,789.99
   Poloniex: 	8,776.76 / 8,791.00
   ----------------------------
   Bittrex/Kraken:	-0.03% [target  0.10%, min -0.42%, max  0.13%]
   Binance/Kraken:	 0.00% [target  0.10%, min -0.24%, max  0.29%]

 [02/12/2018 13:58:15 IN MARKET: Long GDAX / Short Kraken ]
   GDAX/Kraken:	 0.24% [target -0.48%, min  0.00%, max  0.46%]

[ 02/12/2018 13:58:15 IN MARKET: Long Poloniex / Short Kraken ]
   Poloniex/Kraken:	 0.16% [target -0.37%, min -0.09%, max  0.28%]
```
Problems start to occur if you enable or disable exchanges with orders open though - working on it.

Might be ready for PR... soonish? I definitely want to add cancel/resubmits to increase frequency + stop getting caught on legs that don't clear.

**Comment by osheratias** [comment]
Just thinking, what will happen if the bot will open both long and short positions on the same exchange with this implementation?
Could it be a problem?

**Comment by davidjirovec** [comment]
This should be possible, but your USD on exchange should be >= USD spent for long + USD collateral for short + some buffer for short interest & BTC price fluctuations?

**Comment by CCKRK** [comment]
I don't have access to a second exchange to short on right now -- I was thinking of creating a NULL exchange for testing all the possibilities but a lot of things are just going to depend on what information the exchange will give us.

I can't really imagine running into many problems. We should easily stay overall market neutral. I'd be more concerned about hitting that exchanges rate limit. I'm working on that part.

**Comment by CCKRK** [comment]
I am closing this issue as parallel trade functionality has been added to the Parallelism branch. Check it out @davidjirovec but be careful. There's a lot of work to be done.

**Comment by davidjirovec** [comment]
Maybe this should stay open until parallelism is implemented and merged into master? :) I was looking in the related pull request, if supported exchanges are following:

- Poloniex - readme says "BTC/USD trading not supported, BTC/USDT margin trading not supported"
- Binance - readme says "Implementation in progress, BTC/USD not supported"
- Bittrex - readme says "Implementation in progress, BTC/USD not supported (coming soon.)"
- Kraken - this one is ok for me
- GDAX - not supported BTC/USD for me

What currency do you use to test these?

**Comment by CCKRK** [comment]
So I was considering putting exchanges status in a little box in the readme so people could help out, I'll probably do that later today -- I currently use Poloniex+Binance+Bittrex+Kraken+GDAX with USDT as a proxy for USD (the code actually converts it automatically right now -- I should probably add a note on that too.) I added the members to bitfinex and bitstamp recently -- but I have no way to test them.

**Comment by CCKRK** [comment]
@davidjirovec I have updated the readme on Parallelism to more accurately reflect the state of changes in the exchanges. If there's an exchange you'd be willing to test that is not ready let me know and I'll code something up.

**Comment by davidjirovec** [comment]
Bitfinex would be nice.

**Comment by davidjirovec** [comment]
Regarding parallelism I was thinking about possible improvements. How about this:

When entered market with multiple spreads at once, allow exit market with different combinations of exchanges.

For example:
Current status with parallelism - enter market longA/shortB, enter market longC/shortD, exit market longC/shortD, exit market longA/shortB

What I am thinking about - enter market longA/shortB, enter market longC/shortD, exit market longA/shortD, exit market longC/shortB.

This would allow Blackbird to exit market earlier.

**Comment by butor** [comment]
What would be the advantage of doing this? The problem is that if you don't trade A and B at the same moment both for entry and exit, then you will be exposed to market risk.

**Comment by davidjirovec** [comment]
Advantage might be ability to exit market earlier, so you can enter it again earlier.

I am not sure, are you still exposed to market risk if on A, B, C, D you have same amount of BTC?

**Comment by butor** [comment]
I see your thoughts: what you're saying is that the orders for A, B, C and D **all** have the same BTC amounts, right?

In an ideal world yes, you would be hedged against market risk. But the problem is that the BTC amounts could be different: prices not the same between A and C, slippage, fees, etc. Your market risk will be limited, but you wouldn't be totally market neutral.

**Comment by CCKRK** [comment]
Maybe I just don't understand but I'm having trouble imagining a scenario where this would work.

Bitfinex has an appropriate getActivePos and is ready for testing.

**Comment by CCKRK** [comment]
Reopening as it's an active discussion but it'd be easier to track if individual ideas/issues had their own issue
